src/State/Scope.php

```diff
@@ -0,0 +1,310 @@
+<?php
+
+/*
+ * This file is part of Raven.
+ *
+ * (c) Sentry Team
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sentry\State;
+
+use Sentry\Breadcrumbs\Breadcrumb;
+use Sentry\Context\Context;
+use Sentry\Context\TagsContext;
+use Sentry\Context\UserContext;
+use Sentry\Event;
+use Sentry\Severity;
+
+/**
+ * The scope holds data that should implicitly be sent with Sentry events. It
+ * can hold context data, extra parameters, level overrides, fingerprints etc.
+ */
+final class Scope
+{
+    /**
+     * @var Breadcrumb[] The list of breadcrumbs recorded in this scope
+     */
+    private $breadcrumbs = [];
+
+    /**
+     * @var UserContext The user data associated to this scope
+     */
+    private $user;
+
+    /**
+     * @var TagsContext The list of tags associated to this scope
+     */
+    private $tags;
+
+    /**
+     * @var Context A set of extra data associated to this scope
+     */
+    private $extra;
+
+    /**
+     * @var string[] List of fingerprints used to group events together in
+     *               Sentry
+     */
+    private $fingerprint = [];
+
+    /**
+     * @var Severity|null The severity to associate to the events captured in
+     *                    this scope
+     */
+    private $level;
+
+    /**
+     * @var callable[] List of event processors
+     */
+    private $eventProcessors = [];
+
+    /**
+     * Constructor.
+     */
+    public function __construct()
+    {
+        $this->user = new UserContext();
+        $this->tags = new TagsContext();
+        $this->extra = new Context();
+    }
+
+    /**
+     * Sets a new tag in the tags context.
+     *
+     * @param string $key   The key that uniquely identifies the tag
+     * @param string $value The value
+     *
+     * @return $this
+     */
+    public function setTag(string $key, string $value): self
+    {
+        $this->tags[$key] = $value;
+
+        return $this;
+    }
+
+    /**
+     * Gets the tags contained in the tags context.
+     *
+     * @return array<string, string>
+     */
+    public function getTags(): array
```

Comment:
This applies to all `getters` in the `Scope`. I on purpose did not expose the `getters` since we do not want the user to hold on to references of the `context` and modifies them outside of `configureScope` or `withScope`.
Also there should actually never be a reason for the user the call a `getter` of the `Scope`. Again for this reason we have `beforeSend` and `EventProcessors` / `Integrations`.

Reply:
> This applies to all `getters` in the `Scope`. I on purpose did not expose the `getters` since we do not want the user to hold on to references of the `context` and modifies them outside of `configureScope` or `withScope`.

It seems legit. The only downside is that the context objects provide a nice API to manipulate the data while here you can't for example remove a value or merge existing data recursively

> Also there should actually never be a reason for the user the call a `getter` of the `Scope`. Again for this reason we have `beforeSend` and `EventProcessors` / `Integrations`.

I exposed them because I saw the unified API on the Javascript package does the same thing. If the purpose of this new API is to be the same across all SDKs we should have the same methods

Reply:
If someone wants to work with the context they can still modify it in `beforeSend` of an `EventProcessor` directly on the event.

That they are public in JS is also just because of the tests, we will remove the `getters` (that was a mistake on our side).

Reply:
😬 https://github.com/getsentry/sentry-javascript/pull/1671/files#diff-9149adc4b7fd7b844f7874022e9d197d

Reply:
>That they are public in JS is also just because of the tests, we will remove the getters (that was a mistake on our side).

They are useful for testing too in this package, otherwise we have to rely on reflection (the `assertAttribute* methods of PHPUnit are going [to be removed](https://github.com/sebastianbergmann/phpunit/issues/3339) so I would not rely on them anymore)

Reply:
Then we at least should document them with `@internal`.